tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -125,48 +125,107 @@ ROOT::Experimental::Internal::RPageSinkFile::CommitSealedPageImpl(DescriptorId_t
    return WriteSealedPage(sealedPage, bytesPacked);
 }
 
+void ROOT::Experimental::Internal::RPageSinkFile::CommitBatchOfPages(CommittedBatch &batch,
+                                                                     std::vector<RNTupleLocator> &locators)
+{
+   Detail::RNTupleAtomicTimer timer(fCounters->fTimeWallWrite, fCounters->fTimeCpuWrite);
+
+   std::uint64_t offset = fWriter->ReserveBlob(batch.fSize, batch.fBytesPacked);
+
+   auto rangeBegin = batch.fBegin.first;
+   auto rangeEnd = batch.fEnd.first;
+   auto firstPageBegin = batch.fBegin.second;
+   auto lastPageEnd = batch.fEnd.second;
+
+   for (auto rangeIt = rangeBegin; rangeIt != rangeEnd; ++rangeIt) {
+      auto pageBegin = (rangeIt == rangeBegin) ? firstPageBegin : rangeIt->fFirst;
+      auto pageEnd = (std::next(rangeIt) == rangeEnd) ? lastPageEnd : rangeIt->fLast;
+
+      locators.reserve(locators.size() + std::distance(pageBegin, pageEnd));
+
+      for (auto sealedPageIt = pageBegin; sealedPageIt != pageEnd; ++sealedPageIt) {
+         fWriter->WriteIntoReservedBlob(sealedPageIt->GetBuffer(), sealedPageIt->GetBufferSize(), offset);
+         RNTupleLocator locator;
+         locator.fPosition = offset;
+         locator.fBytesOnStorage = sealedPageIt->GetDataSize();
+         locators.push_back(locator);
+         offset += sealedPageIt->GetBufferSize();
+      }
+   }
+
+   fCounters->fNPageCommitted.Add(locators.size());
```

Comment:
Aren't we double counting here, i.e. should be add by `std::distance(pageBegin, pageEnd)` instead?